### ISSUE TITLE
Add more options about writing metadata

### ISSUE THREAD

**Issue by Nakaner** [osmcode/libosmium#241]
Osmium has currently only one option which decides if metadata should be written or not: `add_metadata=true/false` It is supported by `XMLOutputFormat`, `PBFOutputFormat`, `OPLOutputFormat` and `DebugOutputFormat`. `PBFOutputFormat` also [supports](https://github.com/osmcode/libosmium/blob/master/include/osmium/io/detail/pbf_output_format.hpp#L490) the [undocumented](http://osmcode.org/file-formats-manual/) `pbf_add_metadata`.

Geofabrik currently evaluates its download service in regard with privacy regulation which seems to force us to remove some metadata from our public downloads: user names, user IDs and changeset IDs. (These information would still be available to authenticated users) Osmium currently only supports an all or nothing approach. That's why I would like to start a discussion which options should be added.

My initial idea:

* add an `only_version` option which prohibits writing of all metadata except the version.
* add a `no_personal` option which will only write version numbers and timestamps.

However, these new options introduce the need to do some kind of validation. Where should this validation happen? I think that `osmium::io::File::check()` would be a good location (maybe moved to a separate method if the checks become too long).

@joto What do you think?

PS I am currently testing some common OSM related tools if they can work with OSM files which contain only some metadata (no metadata vs. only version numbers vs. version numbers + timestamps). These results will be published soon but at a more appropriate location.

**Comment by joto** [comment]
`pbf_add_metadata` is deprecated in favour of the more general `add_metadata`, that's why it isn't documented.

I don't like the `only_version` and `no_personal` idea, it adds too many options. What if somebody else comes with a different need for other combinations of metadata that they want? We could relatively easily extend the `add_metadata` option to take arguments like `add_metadata=version` (which would only add the version, but not the rest), or `add_metadata=version,changeset_id` (which would have only those two) etc. The special values `true` and `false` would be kept and basically mean "all" and "none".

One thing we have to look into, is the question of whether the PBF file actually supports leaving off only some but not all of these attributes and what software reading them will see. For XML it is quite clear that it would be possible to just not add some attributes, not sure at the moment how flexible the PBF format is. In any case this might lead to problems when reading those files and we have to carefully consider this. Most likely the applications will just get "zero" values for those attributes that are not set and it wont be a big deal, but we have to think through the ramifications here.

@Nakaner What do you mean when you are talking about validation? What validation is needed?

**Comment by Nakaner** [comment]
@joto wrote:
> I don't like the only_version and no_personal idea, it adds too many options. What if somebody else comes with a different need for other combinations of metadata that they want? We could relatively easily extend the add_metadata option to take arguments like add_metadata=version (which would only add the version, but not the rest), or add_metadata=version,changeset_id (which would have only those two) etc. The special values true and false would be kept and basically mean "all" and "none".

I agree. Your idea is better.

> @Nakaner What do you mean when you are talking about validation? What validation is needed?

The options I initially proposed were quite difficult to understand and it is likely that users supply contrary options (`add_metadata=false` + `only_version=true`). That problem does not exist that much with your proposal but it might still be a good idea to check if the provided options are valid and throw an exception if `add_metadata=user_iduser_name` (no comma) is given.

**Comment by Nakaner** [comment]
@joto wrote:
> One thing we have to look into, is the question of whether the PBF file actually supports leaving off only some but not all of these attributes and what software reading them will see. For XML it is quite clear that it would be possible to just not add some attributes, not sure at the moment how flexible the PBF format is. In any case this might lead to problems when reading those files and we have to carefully consider this. Most likely the applications will just get "zero" values for those attributes that are not set and it wont be a big deal, but we have to think through the ramifications here.

I had a look into the format today. All metadata fields of the [`Info` message](https://github.com/scrosby/OSM-binary/blob/master/src/osmformat.proto#L136) are optional. All metadata fields of the [`DenseInfo` message](https://github.com/scrosby/OSM-binary/blob/master/src/osmformat.proto#L155) which is used by the `DenseNodes` message are `repeated`. `repeated` does not mean `required`. To cite the [documentation](https://developers.google.com/protocol-buffers/docs/proto#specifying-field-rules):
> this field can be repeated any number of times (including zero) in a well-formed message. The order of the repeated values will be preserved.

However, Osmium checks if these fields are available [here](https://github.com/osmcode/libosmium/blob/master/include/osmium/io/detail/pbf_decoder.hpp#L626) and throws an exception of type `osmium::pbf_error`.

Here are my results if I write a very simple PBF file using Osmium and try to read it with Osmium, Osmosis and Osmconvert. `only_personal` is a temporary hack in my local copy of Osmium to only write user names, user IDs and changeset IDs (no version and timestamp). `only_non_personal` writes only timestamps and version numbers.

TL;DR The PBF format is not the problem. The applications using it have minor or major bugs.

Osmium fails reading PBF files with DenseNodes and less metadata because it checks it during the reading. Osmosis crashes because it assumes that these files have such metadata and therefore doesn't initialize everything properly. 

The input file:

```xml
<osm version="0.6">
  <node id="1" user="foo" uid="2" changeset="3" timestamp="2017-01-01T00:00:00Z" version="4" lat="50.0" lon="10.0" />
</osm>
```

* written with `osmium cat -o out.osm.pbf --output-format pbf,pbf_dense_nodes=true,pbf_compression=true,only_personal=true testfile.osm`:
  * Osmium throws `osmium::pbf_error` (see above)
  * Osmconvert doesn't read any metadata
  * Osmosis dies with an `IndexOutOfBoundsException` while calling `osmbinary.Osmformat$DenseInfo.getTimestamp()`
* written with `osmium cat -o out.osm.pbf --output-format pbf,pbf_dense_nodes=true,pbf_compression=false,only_personal=true testfile.osm`:
  * Osmium throws `osmium::pbf_error` (see above)
  * Osmconvert fails with `block raw size expected at 0x0A`
  * Osmosis dies with an `IndexOutOfBoundsException` while calling `osmbinary.Osmformat$DenseInfo.getTimestamp()`
* written with `osmium cat -o out.osm.pbf --output-format pbf,pbf_dense_nodes=false,pbf_compression=true,only_personal=true testfile.osm`:
  * Osmium works as expected: `<node id="1" uid="2" user="foo" changeset="3" lat="50" lon="10"/>`
  * Osmconvert does only support DenseNodes (but it supports files without DenseNodes which have all metadata)
  * Osmosis works as expected: `<node id="1" version="-1" timestamp="1969-12-31T23:59:59Z" uid="2" user="foo" changeset="3" lat="50" lon="10"/>`
* written with `osmium cat -o out.osm.pbf --output-format pbf,pbf_dense_nodes=false,pbf_compression=false,only_personal=true testfile.osm`: see test with compression

The same test if I write only timestamp and version number:

* written with `osmium cat -o out.osm.pbf --output-format pbf,pbf_dense_nodes=true,pbf_compression=true,only_non_personal=true testfile.osm`:
  * Osmium throws `osmium::pbf_error` (see above)
  * Osmconvert returns no metadata at all `<node id="1" lat="50.0" lon="10.0"/>`
  * Osmosis dies with an `IndexOutOfBoundsException` while calling `osmbinary.Osmformat$DenseInfo.getUid()`
* written with `osmium cat -o out.osm.pbf --output-format pbf,pbf_dense_nodes=false,pbf_compression=true,only_non_personal=true testfile.osm`:
  * Osmium works as expected: `<node id="1" version="4" timestamp="2017-01-01T00:00:00Z" lat="50" lon="10"/>`
  * Osmconvert fails with `block raw size expected at 0x0A`
  * Osmosis works as expected (same as Osmium)
* The results for the variants with `pbf_compression=false` are the same (except a different error message by Osmconvert)

**Comment by joto** [comment]
See work-in-progress pull request at #242.

**Comment by mmd-osm** [comment]
@Nakaner: leaving out parts of the metadata seems to cause quite some issues, and given that people might be unable to update immediately, some widespread breakage of tools could happen. Did you also evaluate if mapping all personally identifiable metadata to a generic dummy user would be feasible as an intermediate step?

**Comment by joto** [comment]
I have now merged this into master. I am leaving this issue open for the time being until we had the chance to test this some more.

**Comment by Nakaner** [comment]
This issue was fixed by various pull requests to Libosmium and Osmium Tool
